clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -196,7 +202,9 @@ public RecordAppendResult append(TopicPartition tp,
                     return appendResult;
                 }
 
-                MemoryRecordsBuilder recordsBuilder = MemoryRecords.builder(buffer, compression, TimestampType.CREATE_TIME, this.batchSize);
+                byte maxUsableMagic = apiVersions.maxUsableProduceMagic();
+                MemoryRecordsBuilder recordsBuilder = MemoryRecords.builder(buffer, maxUsableMagic, compression,
+                        TimestampType.CREATE_TIME, this.batchSize);
```

Comment:
Good catch. It seems we need to use the same value throughout this logic. We always have the code to handle conversion later if we need it.